core/common/src/main/java/alluxio/underfs/NonAtomicCreateUnderFileSystem.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.underfs;
+
+import alluxio.AlluxioURI;
+import alluxio.Constants;
+import alluxio.exception.ExceptionMessage;
+import alluxio.security.authorization.Permission;
+import alluxio.underfs.options.CreateOptions;
+import alluxio.underfs.options.NonAtomicCreateOptions;
+import alluxio.util.IdUtils;
+import alluxio.util.io.PathUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.OutputStream;
+
+/**
+ * An {@link UnderFileSystem} which does not have an atomic create operation.
+ */
+public abstract class NonAtomicCreateUnderFileSystem extends UnderFileSystem {
```

Comment:
Let's do an in person review for this part. I think providing the default implementation and removing the NonAtomicCreateUnderFileSystem class will mean that a createNonAtomic method or the like will have to be added to object stores (which out-number file systems for the moment).